File: 0275-h-index-ii/0275-h-index-ii.js
function hIndex(citations) {
    const n = citations.length;

    if (n === 0) {
        return 0;
    }

    let left = 0;
    let right = n - 1;
    let firstTrueIndex = -1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);

        // citations[mid]가 충분한 인용 횟수를 가지고 있는지 확인
        if (citations[mid] >= n - mid) {
            firstTrueIndex = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    // h-index 반환
    return firstTrueIndex !== -1
        ? n - firstTrueIndex
        : 0;
}

File: leetcode/0275-h-index-ii.java
class Solution {
    /**
     * Calculates the H-index using the binary search template.
     * Feasible condition: citations[mid] >= n - mid
     */
    public int hIndex(int[] citations) {
        int n = citations.length;
        if (n == 0) {
            return 0;
        }

        // Binary search template
        int left = 0;
        int right = n - 1;
        int firstTrueIndex = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            // Feasible: does paper at mid have enough citations?
            if (citations[mid] >= n - mid) {
                firstTrueIndex = mid;
                right = mid - 1;  // Search for earlier feasible position
            } else {
                left = mid + 1;
            }
        }

        // Return h-index
        return firstTrueIndex != -1 ? n - firstTrueIndex : 0;
    }
}
